pkg/virt-handler/rest/console.go

```diff
@@ -0,0 +1,98 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package rest
+
+import (
+	"net"
+	"net/http"
+	"path"
+	"strconv"
+
+	"github.com/emicklei/go-restful"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+type Console struct {
+	podIsolationDetector isolation.PodIsolationDetector
+}
+
+func NewConsoleResource(podIsolationDetector isolation.PodIsolationDetector) *Console {
+	return &Console{podIsolationDetector: podIsolationDetector}
+}
+
+func (t *Console) VNC(request *restful.Request, response *restful.Response) {
+	stream(request, response, t.getUnixSocketPath(request.PathParameter("uid"), "virt-vnc"))
+}
+
+func (t *Console) Serial(request *restful.Request, response *restful.Response) {
+	stream(request, response, t.getUnixSocketPath(request.PathParameter("uid"), "virt-serial0"))
+}
+
+func (t *Console) getUnixSocketPath(vmiID, socketName string) string {
+	pid, _ := t.podIsolationDetector.GetPid(vmiID)
+	return path.Join("proc", strconv.Itoa(pid), "root", "var", "run", "kubevirt-private", vmiID, socketName)
```

Comment:
I think it would make sense to change the structure of these directories in the pod and use namespace/name. In my opinion, being able to include the namespace+name in the logs to ease debugging would be the main benefit of this.

But I would prefer not to do it without changing the structure of the directories so if we go in this direction, let's do it separately.

Reply:
>I think it would make sense to change the structure of these directories in the pod and use namespace/name.

yep, Since the kubevirt-private dir is completely isolated within the pod, i'd be fine with changing the directory tree to use namespace/name and making the REST endpoint use namespace/name as well. It just requires a little more work here. 

> But I would prefer not to do it without changing the structure of the directories so if we go in this direction, let's do it separately.

what does "separately" mean here? separate PRs?

Reply:
yes, as a separate PR - this one gets too big. I'll post another one for making the port of the console endpoint (8185) configurable so I can do this as part of that PR

Reply:
> @rmohr this is why I recommended using uid in the rest call instead of namespace/name. The uid maps directly to the vnc/console socket's directory path.

I would prefer if we pass in namespace/name and use all already prepared flows which we use in the controller function. They expect the VMI object, which can be easily retrieved from the cache. There is nothing wrong with the UID per se, I would prefer the more high-level and consistent  entrypoint via VMI.

Reply:
For instance, if we pass the VMI to our helpers, it does not matter if we refactor directory structure or not. You would always pass the VMI to the helpers and get a path back. You don't care about the actual form of the path.

Reply:
> For instance, if we pass the VMI to our helpers, it does not matter if we refactor directory structure or not. You would always pass the VMI to the helpers and get a path back. You don't care about the actual form of the path.

Well, but on the other hand it contradicts what others called `principle of least privilege` - a function that seem to require only the UID of the VMI will get a whole VMI. Some of the downsides are that within this function anyone that introduces new code would have access to all the information of the VMI and this function becomes less reusable since it requires the caller to have the VMI at hand.

Yes, it is often that case that such modeling questions involve trade offs. I agree that you pointed out a benefit of passing the whole VMI.

I would prefer to split the `Detect` into smaller functions that require only what they actually need and avoid introducing a VMI informer into virt-handler just for that. But as this can lead to an endless discussion affected by programming style, experience and prioritization I suggest we'll break it here and I'll go with your suggestion :)

Reply:
It is just a layered approach rest-api -> additional checks on the VMI (does it exist, does it have, does it have vnc configured (vmis may not have a vnc device, I don't expect that you directly go to the file and then guess that there may be no vnc device vased on a non-existing file, and so on.)

This is not about style on low-level functions, it is just a normal layered model which allows more checks and decissions to be done on the right levels where you have the informations you need at hand.

Reply:
ok :)